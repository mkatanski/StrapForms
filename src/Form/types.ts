import { ReactNode } from 'react';
import { IComponentProps } from '../common'
import { TInputState } from '../Input'

export interface IFormProps extends IComponentProps {
  children: any,
  componentRenderer?: (props: TFormRendererProps) => ReactNode
}

export type TFormRendererProps = {
  children: ReactNode
  className: string,
  handleFormSubmit: () => {}
}

export type TFormOptions = {
  isPristine: boolean
  isSubmitted: boolean
  isTouched: boolean
  isValid: boolean
}

export type TFormChildContextTypes = {
  registerInput: (inputName: string, data: TInputState) => void,
  deregisterInput: (inputName: string) => void,
}

export type TInputsData = {
  [key: string]: TInputState
}
