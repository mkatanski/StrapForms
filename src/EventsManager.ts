export interface IEventsManager {
  addListener: (eventName: string, cb: TEventListenerMethod) => void
  removeListener: (eventName: string) => void
  clearListeners: () => void
  dispatchEvent: (methodName: string, eData: IEventData) => void
}

export interface IEventData {
  methodArguments: any[]
}

export type TEventListenerMethod = (eventData: IEventData) => void

export type TListenerObject = {
  [key: string]: Array<TEventListenerMethod>,
}

export class EventData implements IEventData {
  private _args: any[]

  constructor(args: any[]) {
    this._args = args
  }

  public get methodArguments(): any[] {
    return this._args
  }
}

export class EventDataAfter extends EventData implements IEventData {
  private _result: any

  constructor(args: any[], result: any) {
    super(args)
    this._result = result
  }

  public get Result(): any {
    return this._result
  }
}

export class EventsManager implements IEventsManager {
  private _listeners: TListenerObject = {}

  public addListener(eventName: string, cb: TEventListenerMethod): void {
    if (!this._listeners[eventName]) {
      this._listeners[eventName] = []
    }
    this._listeners[eventName].push(cb)
  }

  public removeListener(eventName: string): void {
    delete this._listeners[eventName]
  }

  public clearListeners(): void {
    this._listeners = {}
  }

  public dispatchEvent(methodName: string, eData: IEventData): void {
    let listeners

    if (eData instanceof EventData) {
      listeners = this._listeners[`on_before_${methodName}`]
    }

    if (eData instanceof EventDataAfter) {
      listeners = this._listeners[`on_after_${methodName}`]
    }

    if (!listeners) return

    listeners.forEach((listenerMethod) => {
      listenerMethod(eData)
      console.log(eData)
    })
  }
}

export function EventEmitter(target: any, propertyKey: string, descriptor: PropertyDescriptor) {
  var originalMethod = descriptor.value;

  descriptor.value = function (...args: any[]) {
    this.eventsManager && this.eventsManager.dispatchEvent(propertyKey, new EventData(args))
    const result = originalMethod.apply(this, args)
    this.eventsManager && this.eventsManager.dispatchEvent(propertyKey, new EventDataAfter(args, result))

    return result
  }

  return descriptor;
}

