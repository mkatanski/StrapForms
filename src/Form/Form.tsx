import * as React from 'react'
import * as PropTypes from 'prop-types'
import { TInputState } from '../Input'
import { ReactNode } from 'react';

import {
  IFormProps,
  TFormChildContextTypes,
  TFormOptions,
  TFormRendererProps,
  TInputsData
} from './types'

const { Consumer, Provider } = React.createContext({
  registerInput: (inputName: string, inputState: TInputState) => { },
  eventsObservable: {},
});

export const FormContextConsumer = Consumer

export class Form extends React.Component<IFormProps> {
  private inputsData: TInputsData = {}

  private _options: TFormOptions = {
    isPristine: true,
    isSubmitted: false,
    isTouched: false,
    isValid: false,
  }

  private inputs: {}
  private errors: {}
  private warnings: {}

  private eventsObservable: Object = {}

  private _rendererProps: TFormRendererProps = {
    children: this.props.children,
    className: this.props.className,
    handleFormSubmit: this.handleFormSubmit,
  }

  static defaultProps: IFormProps = {
    children: {},
    name: '',
  }

  // componentDidMount() {
    // console.log(this._validationManager)
  // }

  private registerInput(inputName: string, inputState: TInputState) {
    if (!this.inputsData) this.inputsData = {}
    this.inputsData[inputName] = inputState
  }

  // private deregisterInput(inputName: string) {
    // delete this.inputsData[inputName]
    // if (this._validationManager) {
    //   this._validationManager.removeValidator(inputName)
    // }
  // }

  private async handleFormSubmit(): Promise<any> {

  }

  private defaultRenderer(props: TFormRendererProps): ReactNode {
    return (
      <form
        className={props.className}
        onSubmit={props.handleFormSubmit}
      >
        {props.children}
      </form>
    )
  }

  public render(): ReactNode {
    const { componentRenderer, children } = this.props
    return (
      <Provider value={{
        registerInput: this.registerInput.bind(this),
        eventsObservable: this.eventsObservable,
      }}>
        {!!componentRenderer ?
            componentRenderer(this._rendererProps) :
            this.defaultRenderer(this._rendererProps)}
      </Provider>
    )
  }
}


