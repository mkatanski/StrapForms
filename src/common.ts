import { ReactNode } from 'react';

export interface IComponentProps {
  className?: string
  name: string
  key?: any
}
