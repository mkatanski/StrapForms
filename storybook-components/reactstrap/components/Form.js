import React from 'react'
import { StrapForm, StrapFormPropTypes } from '../../../index'

const FormPure = ({ children, handleSubmit }) => (
  <form onSubmit={handleSubmit} >{children}</form>
)

FormPure.propTypes = {
  ...StrapFormPropTypes,
}

export default StrapForm(FormPure)
