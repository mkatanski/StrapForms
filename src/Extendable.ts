import { IEventsManager, EventsManager } from './EventsManager'

export interface IExtendable {
  eventsManager: IEventsManager
}

export type TExtensions = {
  eventsManager?: IEventsManager
}

export class Extendable implements IExtendable {
  private _settings: TExtensions

  constructor(settings?: TExtensions) {
    if (settings) this._settings = settings
  }

  public get eventsManager(): IEventsManager {
    return this._settings ? this._settings.eventsManager : null
  }
}
