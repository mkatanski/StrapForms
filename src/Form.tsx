import * as React from 'react'
import * as PropTypes from 'prop-types'
import { TInputState } from './Input'
import { ReactNode } from 'react';
import { IComponentProps } from './common'

export const { Consumer, Provider } = React.createContext({
  registerInput: (inputName: string, inputState: TInputState) => { }
});

export interface IFormProps extends IComponentProps {
  children: any,
  componentRenderer?: (props: TFormRendererProps) => ReactNode
}

export type TFormRendererProps = {
  children: ReactNode
  className: string,
  handleFormSubmit: () => {}
}

export type TFormOptions = {
  isPristine: boolean
  isSubmitted: boolean
  isTouched: boolean
  isValid: boolean
}

export type TFormChildContextTypes = {
  registerInput: (inputName: string, data: TInputState) => void,
  deregisterInput: (inputName: string) => void,
}

export type TInputsData = {
  [key: string]: TInputState
}

export class Form extends React.Component<IFormProps> {
  private inputsData: TInputsData = {}
  private _options: TFormOptions = {
    isPristine: true,
    isSubmitted: false,
    isTouched: false,
    isValid: false,
  }

  private _rendererProps: TFormRendererProps = {
    children: this.props.children,
    className: this.props.className,
    handleFormSubmit: this.handleFormSubmit,
  }

  static defaultProps: IFormProps = {
    children: {},
    name: '',
  }

  // componentDidMount() {
    // console.log(this._validationManager)
  // }

  private registerInput(inputName: string, inputState: TInputState) {
    if (!this.inputsData) this.inputsData = {}
    this.inputsData[inputName] = inputState
  }

  // private deregisterInput(inputName: string) {
    // delete this.inputsData[inputName]
    // if (this._validationManager) {
    //   this._validationManager.removeValidator(inputName)
    // }
  // }

  private async handleFormSubmit(): Promise<any> {

  }

  private defaultRenderer(props: TFormRendererProps): ReactNode {
    return (
      <form
        className={props.className}
        onSubmit={props.handleFormSubmit}
      >
        {props.children}
      </form>
    )
  }

  public render(): ReactNode {
    const { componentRenderer, children } = this.props
    return (
      <Provider value={{
        registerInput: this.registerInput.bind(this)
      }}>
        {!!componentRenderer ?
            componentRenderer(this._rendererProps) :
            this.defaultRenderer(this._rendererProps)}
      </Provider>
    )
  }
}
