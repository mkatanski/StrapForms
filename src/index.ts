// import { Form } from './Form/Form'
import { Input } from './Input';
import { Form } from './Form/Form';

import { StrapValidator } from './Validation/StrapValidator'

export {
  Form,
  Input,
  // IValidationClass,
  StrapValidator
}
