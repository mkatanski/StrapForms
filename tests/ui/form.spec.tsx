declare let window: any;
declare let global: any, jest: any, jasmine: any, describe: any, clearTest: any, it: any, expect: any, beforeAll: any;
if (typeof jasmine !== 'undefined') jasmine.DEFAULT_TIMEOUT_INTERVAL = 5000;

import { configure } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import * as TestRenderer from 'react-test-renderer';
configure({ adapter: new Adapter() });

import * as React from 'react';
import * as ReactDOM from "react-dom";
import * as enzyme from 'enzyme';

import { Form, Input } from '../../src/';
import { Consumer } from '../../src/Form'

jest.mock('react-dom');

describe('Form', () => {
  it('renders with default component and children', () => {
    const component = TestRenderer.create((
      <Form name='test form'>
        <div>some children component</div>
      </Form>
    ));
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('accepts inputs as children and call registerInput', () => {
    const registerSpy = jest.spyOn(Form.prototype, 'registerInput')
    const component = TestRenderer.create((
      <Form name='test form with input'>
          <Input name='input1' value='test' />
      </Form>
    ));

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    expect(registerSpy).toHaveBeenCalled();
    expect(registerSpy.mock.calls.length).toBe(1);
    expect(registerSpy).toBeCalledWith('input1', {
      "isPristine": true,
      "isTouched": false,
      "isValid": false,
      "value": "test",
    });
  })

  it('renders with custom component and calls registerInput', () => {
    const registerSpy = jest.spyOn(Form.prototype, 'registerInput')
    const component = TestRenderer.create((
      <Form
        name='customComponent'
        componentRenderer={props => {
          return (
            <div>
              {props.children}
            </div>
          )
        }}
      >
          <Input name='input2' />
      </Form>
    ));

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    expect(registerSpy).toHaveBeenCalled();
    expect(registerSpy.mock.calls.length).toBe(2);
    expect(registerSpy).toBeCalledWith('input2', {
      "isPristine": true,
      "isTouched": false,
      "isValid": false,
      "value": "",
    });
  })
});
