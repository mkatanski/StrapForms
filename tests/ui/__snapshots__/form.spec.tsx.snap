// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form accepts inputs as children and call registerInput 1`] = `
<form
  className={undefined}
  onSubmit={[Function]}
>
  <input
    className={undefined}
    name="input1"
    onBlur={[Function]}
    onChange={[Function]}
    value="test"
  />
</form>
`;

exports[`Form renders with custom component and calls registerInput 1`] = `
<div>
  <input
    className={undefined}
    name="input2"
    onBlur={[Function]}
    onChange={[Function]}
    value=""
  />
</div>
`;

exports[`Form renders with default component and children 1`] = `
<form
  className={undefined}
  onSubmit={[Function]}
>
  <div>
    some children component
  </div>
</form>
`;
